---

# DS400: Assignment I (Python)

##### Mohamed Elsherif (Student ID: 6378374)
##### November 1, 2024


---

## Importing data

<div class="alert alert-block alert-info">
1. Import the data into R and python, respectively. In R, try out all three functions discussed in the lecture: utils:::read.csv(), readr:::read_csv() and data.table:::fread() on at least one of the two tables, and check whether the shape of the imported data is the same in every case. In python, use pandas.read_csv() to import the data.</div>

In [1]:
# Library Setup
import os
import pandas as pd

In [2]:
# Set working directory
os.chdir('C:/Users/Admin/Downloads/DS400_WS24-25/Assignment01')

# Verify working directory
print("Current Working Directory: ", os.getcwd())

Current Working Directory:  C:\Users\Admin\Downloads\DS400_WS24-25\Assignment01


In [3]:
# Load datasets
AirBNB_listings_muc = pd.read_csv("./data/listings_muc.csv")
AirBNB_calendar_muc = pd.read_csv("./data/calendar_muc.csv")

<div class="alert alert-block alert-info">
2. Print the first five rows and column types in R and python, respectively</div>

In [4]:
# Display the first five rows of listing dataset
print(AirBNB_listings_muc.shape)
print(AirBNB_listings_muc.dtypes)
print(AirBNB_listings_muc.isna().sum())
AirBNB_listings_muc.head()

(6746, 16)
id                                float64
name                               object
host_id                             int64
host_name                          object
neighbourhood_group               float64
neighbourhood                      object
latitude                          float64
longitude                         float64
room_type                          object
number_of_reviews                   int64
last_review                        object
reviews_per_month                 float64
calculated_host_listings_count      int64
availability_365                    int64
number_of_reviews_ltm               int64
license                            object
dtype: object
id                                   0
name                                 0
host_id                              0
host_name                            1
neighbourhood_group               6746
neighbourhood                        0
latitude                             0
longitude                     

id                                               name  host_id  \
0   97945.0                   Deluxw-Apartm. with roof terrace   517685   
1  114695.0                 Apartment Munich/East with sundeck   581737   
2  127383.0                  City apartment next to Pinakothek   630556   
3  159634.0  Fancy, bright central roof top flat and homeof...   765694   
4  170154.0              Own floor & bath, parking & breakfast   108297   

  host_name  neighbourhood_group       neighbourhood  latitude  longitude  \
0  Angelika                  NaN              Hadern  48.11492   11.48954   
1   Stephan                  NaN        Berg am Laim  48.12071   11.63758   
2     Sonja                  NaN         Maxvorstadt  48.15199   11.56482   
3    Susana                  NaN  Pasing-Obermenzing  48.13855   11.46586   
4      Alex                  NaN   Sendling-Westpark  48.10814   11.52733   

         room_type  number_of_reviews last_review  reviews_per_month  \
0  Entire home/apt                127  2019-10-03               0.82   
1  Entire home/apt                 69  2024-06-22               0.44   
2  Entire home/apt                112  2024-02-04               0.70   
3  Entire home/apt                 40  2024-05-28               0.26   
4     Private room                559  2024-06-21               3.60   

   calculated_host_listings_count  availability_365  number_of_reviews_ltm  \
0                               2               200                      0   
1                               3               108                     12   
2                               1                 0                      5   
3                               1                12                      4   
4                               1               123                     27   

  license  
0     NaN  
1     NaN  
2     NaN  
3     NaN  
4     NaN

In [5]:
# Display the first few rows of listing availability dates dataset PLUS Data integrity screening
print(AirBNB_calendar_muc.shape)
print(AirBNB_calendar_muc.dtypes)
print(AirBNB_calendar_muc.isna().sum())
AirBNB_calendar_muc.head()

(2462238, 6)
listing_id        float64
date               object
available            bool
price              object
adjusted_price    float64
min_max_nights     object
dtype: object
listing_id              0
date                    0
available               0
price                   0
adjusted_price    2462238
min_max_nights          0
dtype: int64


listing_id        date  available   price  adjusted_price min_max_nights
0     97945.0  2024-06-27      False  $95.00             NaN           2-90
1     97945.0  2024-06-28      False  $95.00             NaN           2-90
2     97945.0  2024-06-29      False  $95.00             NaN           2-90
3     97945.0  2024-06-30      False  $95.00             NaN           2-90
4     97945.0  2024-07-01      False  $95.00             NaN           2-90

## Tidying the data
<div class="alert alert-block alert-info">   
4. Please write down the requirements for some data table to be tidy in your submission notebook.
</div>

<div class="alert alert-block alert-info" style="background-color: yellow;">   
    
**Requirments for data tables to be Tidy:**

- Each variable forms a column.
- Each observation forms a row.
- Each type of observational unit forms a table.
</div>

<div class="alert alert-block alert-info">   
5. Can any of the two tables be considered tidy? If not, what would you have to do to make them tidy?
Please perform all necessary steps to make the data tidy. Print the first five rows of the resulting tables.
</div>

<div class="alert alert-block alert-info" style="background-color: yellow;">   
    
##### I- listings_muc.csv
    
**listings_muc.csv is arguably NOT tidy for the following reasons:**

- <u>**Main Reason**</u> The 'host_id' and 'host_name' columns form a different type of observational unit and are categorical variables that won't change, Therefore we can create a separate table to store this information. This task wasn't explicitly mentioned in the assignment, so I didn't include it as part of my solution below.
- The enitre 'neighbourhood_group' column and major portion of 'license' column have missing values, which questions the purpose of having both variables included for analysis. To rectify this, we should remove both. 
- Missing values in columns 'last_review' and its current values are stored as an object datatype, requiring conversion to datetime for analysis. However since dates of 'last_review' offers limited value in current analysis unless feature engineered (e.g., into "days since last review"), I opted to drop it from the dataset.
- Missing values in column 'reviews_per_month' columns when 'number_of_reviews' has value of 0. To rectify this, we can replace those missing values with 0. 
- The column 'id' needs to be converted from float in to integer or object (to avoid issues with decimals). In R, no need to perform such step as R will automatically remove the '.0'.
- The columns 'name' and 'host_name' (which has 1 missing value) seem irrelevant to analysis and can be removed.</div>

In [6]:
# Drop 'name', 'host_name', 'neighbourhood_group', 'last_review', and 'license' columns
AirBNB_listings_muc.drop(columns=['name', 'host_name', 'neighbourhood_group', 'last_review', 'license'], inplace=True)

# Convert the 'id' column to integer 
AirBNB_listings_muc['id'] = AirBNB_listings_muc['id'].astype('int64')

# Set 'reviews_per_month' to 0 where 'number_of_reviews' is 0
AirBNB_listings_muc.loc[AirBNB_listings_muc['number_of_reviews'] == 0, 'reviews_per_month'] = 0

<div class="alert alert-block alert-info" style="background-color: yellow;">   
    
##### II- calendar_muc.csv
    
**calendar_muc.csv is NOT tidy because:**

- <u>**Main Reason**</u> The 'min_max_nights' information are two different variables placed in one column, and each cell has two values. To rectify this, we should separate into two columns: 'minimum_nights' and 'maximum_nights'.
- The columns 'price' and 'date' need to be converted to float/numeric (after removing the '$' sign) and datetime format respectively.
- The column 'listing_id' need to be converted to integer or object (to avoid issues with decimals). In R, no need to perform such step as R will automatically remove the '.0'.
- The enitre 'adjusted_price' column has missing values, which questions the purpose of having this variable included for analysis. To rectify this, we should remove it.
</div>

In [7]:
# Splitting 'min_max_nights' into 'min_nights' and 'max_nights'
AirBNB_calendar_muc[['minimum_nights', 'maximum_nights']] = AirBNB_calendar_muc['min_max_nights'].str.split('-', expand=True)

# Converting new columns to integers
AirBNB_calendar_muc['minimum_nights'] = AirBNB_calendar_muc['minimum_nights'].astype('int64')
AirBNB_calendar_muc['maximum_nights'] = AirBNB_calendar_muc['maximum_nights'].astype('int64')

# Dropping the original 'min_max_nights' column
AirBNB_calendar_muc = AirBNB_calendar_muc.drop(columns=['min_max_nights'])

# Convert 'price' to numeric by removing the dollar sign and converting to float
AirBNB_calendar_muc['price'] = AirBNB_calendar_muc['price'].replace('[\$,]', '', regex=True).astype(float)

# Convert 'date' to datetime format
AirBNB_calendar_muc['date'] = pd.to_datetime(AirBNB_calendar_muc['date'], format='%Y-%m-%d')

# Convert 'listing_id' to integer 
AirBNB_calendar_muc['listing_id'] = AirBNB_calendar_muc['listing_id'].astype('int64')

# Drop the 'adjusted_price' column as it has no values
AirBNB_calendar_muc = AirBNB_calendar_muc.drop(columns=['adjusted_price'])

<div class="alert alert-block alert-info">   
6. What is the shape of the tables after making the data tidy?</div>

In [8]:
# Verify changes and print shape of listing
print(AirBNB_listings_muc.shape)
print(AirBNB_listings_muc.dtypes)
print(AirBNB_listings_muc.isna().sum())
AirBNB_listings_muc.head(3)

(6746, 11)
id                                  int64
host_id                             int64
neighbourhood                      object
latitude                          float64
longitude                         float64
room_type                          object
number_of_reviews                   int64
reviews_per_month                 float64
calculated_host_listings_count      int64
availability_365                    int64
number_of_reviews_ltm               int64
dtype: object
id                                0
host_id                           0
neighbourhood                     0
latitude                          0
longitude                         0
room_type                         0
number_of_reviews                 0
reviews_per_month                 0
calculated_host_listings_count    0
availability_365                  0
number_of_reviews_ltm             0
dtype: int64


id  host_id neighbourhood  latitude  longitude        room_type  \
0   97945   517685        Hadern  48.11492   11.48954  Entire home/apt   
1  114695   581737  Berg am Laim  48.12071   11.63758  Entire home/apt   
2  127383   630556   Maxvorstadt  48.15199   11.56482  Entire home/apt   

   number_of_reviews  reviews_per_month  calculated_host_listings_count  \
0                127               0.82                               2   
1                 69               0.44                               3   
2                112               0.70                               1   

   availability_365  number_of_reviews_ltm  
0               200                      0  
1               108                     12  
2                 0                      5

In [9]:
# Verify changes and print shape of calendar
print(AirBNB_calendar_muc.shape)
print(AirBNB_calendar_muc.dtypes)
print(AirBNB_calendar_muc.isna().sum())
AirBNB_calendar_muc.head(5)

(2462238, 6)
listing_id                 int64
date              datetime64[ns]
available                   bool
price                    float64
minimum_nights             int64
maximum_nights             int64
dtype: object
listing_id        0
date              0
available         0
price             0
minimum_nights    0
maximum_nights    0
dtype: int64


listing_id       date  available  price  minimum_nights  maximum_nights
0       97945 2024-06-27      False   95.0               2              90
1       97945 2024-06-28      False   95.0               2              90
2       97945 2024-06-29      False   95.0               2              90
3       97945 2024-06-30      False   95.0               2              90
4       97945 2024-07-01      False   95.0               2              90

---

### Merging the data

<div class="alert alert-block alert-info">
7. In the calendar data, change the name of the listing_id column to id.</div>

In [10]:
# Rename 'listing_id' to 'id' in the AirBNB_calendar_muc DataFrame
AirBNB_calendar_muc.rename(columns={'listing_id': 'id'}, inplace=True)

<div class="alert alert-block alert-info">
8.  Aggregate the calendar data to the listing level. The variable available should be transformed into the number of available days per listing. The variable date should be discarded. The variables price, minimum_nights, and maximum_nights should be aggregated to the listing level..</div>

In [11]:
# Aggregate the calendar data
calendar_agg = AirBNB_calendar_muc.groupby('id').agg(
    available=('available', 'sum'),  # Count total available days
    price=('price', 'last'),               # Get the last price for each listing
    minimum_nights=('minimum_nights', 'min'),  # Minimum nights required
    maximum_nights=('maximum_nights', 'max')   # Maximum nights allowed
).reset_index()

<div class="alert alert-block alert-info">
9. Merge the aggregated calendar and the listings data by id. Call the resulting data frame merged and
print its shape.</div>

In [12]:
# Merge with the listings dataset
merged = pd.merge(AirBNB_listings_muc, calendar_agg, on='id', how='left')

print(merged.shape)
merged.head(5)

(6746, 15)


id  host_id       neighbourhood  latitude  longitude        room_type  \
0   97945   517685              Hadern  48.11492   11.48954  Entire home/apt   
1  114695   581737        Berg am Laim  48.12071   11.63758  Entire home/apt   
2  127383   630556         Maxvorstadt  48.15199   11.56482  Entire home/apt   
3  159634   765694  Pasing-Obermenzing  48.13855   11.46586  Entire home/apt   
4  170154   108297   Sendling-Westpark  48.10814   11.52733     Private room   

   number_of_reviews  reviews_per_month  calculated_host_listings_count  \
0                127               0.82                               2   
1                 69               0.44                               3   
2                112               0.70                               1   
3                 40               0.26                               1   
4                559               3.60                               1   

   availability_365  number_of_reviews_ltm  available  price  minimum_nights  \
0               200                      0        200   95.0               2   
1               108                     12        108  105.0               1   
2                 0                      5          0  120.0               3   
3                12                      4         12   55.0               1   
4               123                     27        123   70.0               3   

   maximum_nights  
0              90  
1              30  
2              20  
3            1125  
4            1125

In [13]:
# Check for missing values in each column of the merged DataFrame
missing_values = merged.isna().sum()

# Print the missing values for each column
print("Missing values in each column:")
print(missing_values)

Missing values in each column:
id                                0
host_id                           0
neighbourhood                     0
latitude                          0
longitude                         0
room_type                         0
number_of_reviews                 0
reviews_per_month                 0
calculated_host_listings_count    0
availability_365                  0
number_of_reviews_ltm             0
available                         0
price                             0
minimum_nights                    0
maximum_nights                    0
dtype: int64


---

## Analyzing the data

<div class="alert alert-block alert-info">
10. How many neighborhoods are there in Munich? Print the number of neighborhoods and all their names.
How many and which room types are there?</div>

In [14]:
# Count neighborhoods and print their names
neighborhoods = merged['neighbourhood'].unique()  # Get unique neighborhoods
num_neighborhoods = len(neighborhoods)  # Count unique neighborhoods

print(f"There are {num_neighborhoods} neighborhoods in Munich:")
print(neighborhoods)

# Count room types and print their names
room_types = merged['room_type'].unique()  # Get unique room types
num_room_types = len(room_types)  # Count unique room types

print(f"\nThere are {num_room_types} room types available:")
print(room_types)

There are 25 neighborhoods in Munich:
['Hadern' 'Berg am Laim' 'Maxvorstadt' 'Pasing-Obermenzing'
 'Sendling-Westpark' 'Neuhausen-Nymphenburg' 'Schwabing-West'
 'Allach-Untermenzing' 'Au-Haidhausen' 'Schwanthalerhöhe' 'Bogenhausen'
 'Ludwigsvorstadt-Isarvorstadt' 'Sendling' 'Ramersdorf-Perlach'
 'Thalkirchen-Obersendling-Forstenried-Fürstenried-Solln'
 'Untergiesing-Harlaching' 'Obergiesing' 'Schwabing-Freimann'
 'Tudering-Riem' 'Feldmoching-Hasenbergl' 'Milbertshofen-Am Hart'
 'Altstadt-Lehel' 'Laim' 'Moosach' 'Aubing-Lochhausen-Langwied']

There are 4 room types available:
['Entire home/apt' 'Private room' 'Shared room' 'Hotel room']


<div class="alert alert-block alert-info">
11. Regress the number of available days on the price and the minimum_nights. Interpret the results. What would you say – is this a good model?</div>

In [15]:
import statsmodels.api as sm

# Define the independent variables (price and minimum_nights) and add a constant for the intercept
X = merged[['price', 'minimum_nights']]
X = sm.add_constant(X)       

# Define the dependent variable (available)
y = merged['available']

# Fit the OLS regression model
model = sm.OLS(y, X).fit()
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:              available   R-squared:                       0.005
Model:                            OLS   Adj. R-squared:                  0.005
Method:                 Least Squares   F-statistic:                     17.97
Date:                Sun, 03 Nov 2024   Prob (F-statistic):           1.65e-08
Time:                        15:34:26   Log-Likelihood:                -41729.
No. Observations:                6746   AIC:                         8.346e+04
Df Residuals:                    6743   BIC:                         8.348e+04
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
const            107.0068      1.493     71.

<div class="alert alert-block alert-info" style="background-color: yellow;">   
    
 #### **Interpretation:**

- <u>**Price Coefficient (0.0039)**</u> indicates that for each additional unit (i.e. $) increase in price, the number of available days is expected to increase by approximately 0.0039 days, holding the minimum nights constant. The coefficient is statistically significant (p = 0.001), suggesting that price has a positive effect on availability. However, the effect size is very small.
- <u>**Minimum Nights Coefficient (0.2371)**</u> indicates that for each additional night added to the minimum stay requirement, the number of available days is expected to increase by about 0.2371 days, holding price constant. This coefficient is also statistically significant (p < 0.001), suggesting that there is a positive relationship between minimum nights and availability.

**No the model is not good,** given the very low R-squared value (0.005), as the model explains very little of the variance in the number of available days. 
</div>

<div class="alert alert-block alert-info">
12. In a second model, regress the number of available days on the price and the minimum_nights, but include neighborhood and room type as fixed effects. Interpret the results. Do you see any problems with this model?</div>

In [16]:
import statsmodels.formula.api as smf

# Define the formula
formula = 'available ~ price + minimum_nights + C(neighbourhood) + C(room_type)'

# Fit the regression model
second_model = smf.ols(formula=formula, data=merged).fit()
print(second_model.summary())

                            OLS Regression Results                            
Dep. Variable:              available   R-squared:                       0.034
Model:                            OLS   Adj. R-squared:                  0.030
Method:                 Least Squares   F-statistic:                     8.272
Date:                Sun, 03 Nov 2024   Prob (F-statistic):           2.41e-34
Time:                        15:34:27   Log-Likelihood:                -41629.
No. Observations:                6746   AIC:                         8.332e+04
Df Residuals:                    6716   BIC:                         8.352e+04
Df Model:                          29                                         
Covariance Type:            nonrobust                                         
                                                                                 coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------

<div class="alert alert-block alert-info" style="background-color: yellow;">   
    
#### **Interpretation:**

- <u>**Price Coefficient (0.0044)**</u> indicates that for each additional unit (i.e. $) increase in price, the number of available days is expected to increase by approximately 0.0044 days, holding the minimum nights constant. The coefficient is statistically significant (p < 0.001), suggesting that price has a positive effect on availability. However, the effect size is still very small.
- <u>**Minimum Nights Coefficient (0.2667)**</u> indicates that for each additional night added to the minimum stay requirement, the number of available days is expected to increase by about 0.2667 days, holding price constant. This coefficient is also statistically significant (p < 0.001), suggesting that there is a positive relationship between minimum nights and availability.

- <u>**Neighborhood Effects**</u> shows several neighborhoods having negative coefficients such as Altstadt-Lehel (-37.06 days), Au-Haidhausen (-62.24 days), Bogenhausen (-40.27 day) and Hadern (-60.18 days) indicating fewer available days compared to the baseline neighborhood. These negative coefficients suggest that listings in these neighborhoods tend to have significantly fewer available days compared to the reference neighborhood. Neighborhoods such as Untergiesing-Harlaching and Tudering-Riem have coefficients close to zero, indicating minimal effect on availability.

- <u>**Room Type Effects**</u> shows coefficient 118.43 for Hotel room, suggesting that hotel rooms have significantly more available days than the baseline room type compared to other room types.

**The problems I see in this model** is that the R-squared value (0.034) is still low indicating that the model does not explain much of the variance in the dependent variable and that there may be other important variables not included in this model. The low R-squared value could also be due to presence of multicollinearity or other factors that violate assumptions of OLS which are: normally distributed residuals, homoscedasticity (constant variance of residuals), and independence of observations.
</div>

---
# END OF SOLUTION
---